gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1353 @@
+
+		if (!OidIsValid(nspOid))
+		{
+			elog(DEBUG1, "[tracking_get_track] schema \"%s\" does not exist", name);
+			continue;
+		}
+
+		tf_get_global_state.schema_oids = lappend_oid(tf_get_global_state.schema_oids, nspOid);
+	}
+
+	if (schema_names)
+		pfree(schema_names);
+}
+
+
+static bool
+schema_is_tracked(Oid schema)
+{
+	ListCell   *lc;
+
+	if (tf_get_global_state.schema_oids == NIL)
+		return false;
+
+	foreach(lc, tf_get_global_state.schema_oids)
+	{
+		Oid			tracked_schema = lfirst_oid(lc);
+
+		if (tracked_schema == schema)
+			return true;
+	}
+
+	return false;
+}
+
+static bool
+kind_is_tracked(char type, uint64 allowed_kinds)
+{
+	if (type >= 'a' && type <= 'z')
+		return (allowed_kinds & (1UL << (type - 'a'))) != 0;
+	else
+		return (allowed_kinds & (1UL << (26 + (type - 'A')))) != 0;
+}
+
+/*
+ * Main logic for getting the size track.
+ */
+Datum
+tracking_get_track_main(PG_FUNCTION_ARGS)
+{
+	FuncCallContext *funcctx;
+	tf_main_func_state_t *state;
+	HeapTuple	result;
+	Datum		datums[GET_TRACK_TUPDESC_LEN];
+	bool		nulls[GET_TRACK_TUPDESC_LEN] = {0};
+
+	tf_check_shmem_error();
+
+	if (!pg_atomic_unlocked_test_flag(&tf_shared_state->tracking_error))
+		ereport(ERROR,
+				(errmsg("Can't perform tracking for database %u properly due to internal error", MyDatabaseId)));
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldcontext;
+
+		funcctx = SRF_FIRSTCALL_INIT();
+
+		RegisterXactCallbackOnce(xact_end_get_callback, NULL);
+
+		oldcontext = MemoryContextSwitchTo(CurTransactionContext);
+
+		if (tf_get_global_state.bloom == NULL)
+		{
+			tf_get_global_state.bloom = palloc(FULL_BLOOM_SIZE(bloom_size));
+			if (!bloom_set_move(&tf_shared_state->bloom_set, MyDatabaseId, tf_get_global_state.bloom))
+				ereport(ERROR,
+						(errcode(ERRCODE_GP_COMMAND_ERROR),
+						 errmsg("database %u is not tracked", MyDatabaseId),
+					errhint("Call 'arenadata_toolkit.tracking_register_db()'"
+							"to enable tracking")));
+		}
+		else
+		{
+			/*
+			 * This code is needed for the cases when there are several track
+			 * requests within the same transaction. rollback_bloom stands for
+			 * preserving initial filter state at the moment of the first
+			 * function call within the transaction.
+			 */
+			if (tf_get_global_state.rollback_bloom == NULL)
+			{
+				tf_get_global_state.rollback_bloom = palloc(FULL_BLOOM_SIZE(bloom_size));
+				bloom_copy(tf_get_global_state.bloom, tf_get_global_state.rollback_bloom);
+			}
+			bloom_clear(tf_get_global_state.bloom);
+			if (!bloom_set_move(&tf_shared_state->bloom_set, MyDatabaseId, tf_get_global_state.bloom))
+				ereport(ERROR,
+						(errcode(ERRCODE_GP_COMMAND_ERROR),
+						 errmsg("database %u is not tracked", MyDatabaseId),
+					errhint("Call 'arenadata_toolkit.tracking_register_db()'"
+							"to enable tracking")));
+		}
+		/* initial snapshot shouldn't return drops */
+		if (tf_get_global_state.bloom && !tf_get_global_state.bloom->is_set_all)
```

Comment:
Is it possible for `tf_get_global_state.bloom` to be NULL here? Do we need the NULL check?